Create end-user-q-and-a-04.md (#3041)
open-telemetry/opentelemetry.io: content/en/blog/2023/end-user-q-and-a-04.md

---
title: 'End-User Q&A Series: Migrating to OTel at Lightstep'
linkTitle: 'End-User Q&A: Migrating to OTel at Lightstep'
date: 2023-07-14
author: '[Reese Lee](https://github.com/reese-lee) (New Relic)'

> body_class: otel-with-contributions-from

Reviewer: ```suggestion
body_class: otel-with-contributions-from
spelling: cSpell:ignore Aronoff codepath Boten fluentbit autoscaler kubelet
```

---

With contributions from [Adriana Villela](https://github.com/avillela) 
(Lightstep from ServiceNow).

For the OpenTelemetry (OTel) End User Working Group's fourth 
[End User Q&A session](/community/end-user/interviews-feedback/) of 2023, we 
spoke with [Jacob Aronoff](https://www.linkedin.com/in/jaronoff97), Staff 
Software Engineer at [Lightstep from ServiceNow](https://lightstep.com/) and an 
OpenTelemetry Operator Maintainer. Read on if you are interested in learning 
how a vendor is using OTel in-house! 

This series of interviews is a monthly casual discussion with a team that's 
using OpenTelemetry in production. The goal is to share with the community what 
we've learned about how they are doing this, along with their successes and 
challenges, so that we can help improve OpenTelemetry together. 

## Overview

In this session, Jacob shared:

- How he approached migrating to OpenTelemetry from OpenTracing and OpenCensus
- What the [`TargetAllocator`](https://github.com/open-telemetry/opentelemetry-operator/blob/main/cmd/otel-allocator/README.md) 

> - is, and how he's using it today

Reviewer 2: should these lines be combined?

- Why you might not want to deploy your Collector as a sidecar 

## The interview 

### The backstory

Jacob has been on the Telemetry Pipeline Team at Lightstep from ServiceNow for 
almost two years now. He spent the first year solely focused on OTel migrations, 
internally as well as making it easier for their customers. 

When he joined the team, he says "we were still on OpenTracing for tracing, and 
a mix of OpenCensus and some hand-rolled Statsd stuff for metrics." This meant 
they had to run a proxy on every single Kubernetes pod (where a proxy sits as a 

> sidecar on every pod, which means that you have to run another application that's 

Author: Linking this further down in the post, in the section that talks about the other deployment modes.

"Another thing I had to check was that all the attributes were still present 
before and after migration, which is another thing that matters," Jacob notes. 
Sometimes they weren't, as in the case of Statsd automatically adding something 
they didn't care about; those could be safely ignored.

For tracing, Jacob says, "I picked a service that had both internal-only traces 
(stayed within a single service) and traces that spanned multiple services with 
different types of instrumentation, so from Envoy to OTel to OpenTracing." 

He explains, "What you want to see is that the trace before has the same 
structure as the trace after. So I made another script that checked that those 
structures were relatively the same and that they all had the same attributes as 
well... That’s the point of the tracing migration – what matters is that all the 
attributes stayed the same."

### When data goes missing

"The ‘why it’s missing’ stories are the really complicated ones," says Jacob. 
Sometimes, it's as simple as forgetting "to add something somewhere," but other 
times, there could be an upstream library that doesn't emit what you expected 
for OTel. 

> He tells a story about the time he  migrated their grpc util package (which is 

Reviewer 2: ```suggestion
He tells a story about the time he migrated their grpc util package (which is 
```

> He tells a story about the time he  migrated their grpc util package (which is 

Reviewer 2: [we could probably link out here](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrumentation/google.golang.org/grpc/otelgrpc)

now in Go contrib) and found an issue with propagation. 

"I was trying to understand what’s going wrong here. When I looked at the code – 
this tells you how early I was doing this migration – where there was supposed to 

> be a propagator, there was just a 'TODO' from Alex Boten," he shares. "It just 

Reviewer 2: nit: can we remove alex's name here? no need for the callout 😅

Author: 😂 Yes, absolutely, I thought it was funny @codeboten

available to the community. 

"A lot of the metrics work resulted in big performance boosts for OTel metrics," 
he says. "Like OTel Go metrics. It also has given the Stats folks some ideas 
about how descriptive the API should be for various features. 
So things like Views and the use of Views is something we used heavily early in 
the migration."

### Metrics Views

"A Metrics View is something that is run inside of your Meter Provider in OTel," 
Jacob explains. There are many configuration options, such as dropping 
attributes, which is one of the most common use cases. "For example, you’re a 
centralized SRE and you don't want anyone to instrument code with any user ID 
attribute, because that’s a high cardinality thing and it’s going to explode 
your metrics cost. You can make a View that gets added to your instrumentation 
and tell it to not record it, to deny it."

There are also more advanced use cases, for example, dynamically changing the 
temporality or aggregation of your metrics. Temporality refers to whether a 
metric incorporates the previous measurement or not (cumulative and delta), 
and aggregation refers to how you send off the metrics. 

> "It’s most useful for histograms," says Jacob. "When you record histograms, 

Reviewer 2: i may amend this to say `It's most useful for [our] histograms`, because I don't think a lot of other people will have this issue.

there." He notes that there may have been recent changes to this that he's not 
aware of at this time. 

"For a long time, we’ve used span events and logs for a lot of things 
internally," he says. "I’m a big fan of them." He is not as big a fan of 
logging, sharing that he thinks they are "cumbersome and expensive." He suggests 
that users opt for tracing and trace logs whenever possible, although he does 
like logging for local development, and tracing for distributed development." 

### Telemetry collection in Kubernetes

Kubernetes now has the ability to emit OTel traces natively, and Jacob is 
interested in seeing if the traces they get from those are sufficient for 
generating better Kubernetes metrics using the span-to-metrics processor. 

"I'm very focused on infrastructure metrics, like Kubernetes infrastructure 
metrics, and I find them to be very painful in their current form," he says. 
Currently, he is using the Prometheus APIs to collect them, which is the 
ubiquitous way in the observability community since Kubernetes already emits 
these natively.  

"That's what we do right now, and I use an OTel component that I work on called 

> the target allocator to distribute those targets, which is a pretty efficient way 

Reviewer 2: ```suggestion
the Target Allocator to distribute those targets, which is a pretty efficient way 
```

make sure they work internally first. 

"We're running in a single path where there could be two collectors in two 
environments that could be running two different images and two different versions. 
It gets really meta and really confusing to talk about," he says. "And then, if 
you’re sending Collector A across an environment to Collector B, Collector B also
 emits telemetry about itself, which is then collected by Collector C, so it just 
 chains."

In a nutshell, you need to make sure that the collector is actually working. 
"That’s like the problem when we’re debugging this stuff. When there’s a problem 
you have to think up where the problem actually is -- is it in how we collect the 
data, is it in how we emit the data, is it in the source of how the data was 
generated? One of a bunch of things." 

### Kubernetes modes on OTel

There are four deployment modes for running Kubernetes on OTel: 
[Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/), 
[DaemonSet](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/), 
[StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/), 

> and [Sidecar](). Which ones you should use depends on what you need to do, such 

Reviewer: here's a link missing

as how you like to run applications for reliability. 

"Sidecar is the one we use the least and is probably used the least across the 
industry if I had to make a bet," Jacob says. "They’re expensive. If you don’t 
really need them, then you shouldn’t use them." An example of something run as a 
sidecar is Istio, "which makes a lot of sense to run as a sidecar because it 

> does traffic proxy and it hooks into your container network to change how it 

Reviewer 2: ```suggestion
does proxy traffic and it hooks into your container network to change how it 
```

On the other hand, if you have five pods deployed on StatefulSets, "that's not 
that expensive." When you run in StatefulSet mode, you get an exact number of 
replicas that should exist at all times, each with a predictable name -- which 
is "a really valuable thing when you want consistent IDs." 

Due to the consistent IDs, you can do some extra work with the Target Allocator, 
which is why it's required. Another thing that StatefulSets guarantee is 
something called in-place deployment, which is also available with DaemonSets; 
this is where you take the pod down before you create a new one. 

"In a deployment you usually do a 1-up, 1-down, or what’s called a rolling 
deployment, or rolling update," Jacob says. If you were doing this with the 
Target Allocator, you are likely to get much more unreliable scrapes. This is 
because you have to redistribute all the targets when a new replica comes up, 
because the hash ring you place these on has changed, requiring a recalculation 
of all the hashes you've assigned. 

Whereas with StatefulSets, this isn't necessary, since you get a consistent ID 
range. "So when you do 1-down 1 up, it keeps the same targets each time. So like 

> a placeholder for it – you don’t have to recalculate the ring," he exaplins. 

Reviewer: ```suggestion
a placeholder for it – you don’t have to recalculate the ring," he explains. 
```

> He notes that this is really only useful as a  metrics use case, where you're 

Reviewer 2: ```suggestion
He notes that this is really only useful as a metrics use case, where you're 
```

scraping Prometheus. He notes that they'd probably run it as a Deployment for 
anything else, since that mode gives you most everything you would need. 

> Collectors are stateless, so there is no need for them to hold on to anything, 

Reviewer 2: ```suggestion
Collectors are usually stateless, so there is no need for them to hold on to anything, 
```

and Deployments are leaner as a result. "You can just run and roll out and 
everyone’s happy," he says. "That’s how we run most of our collectors, is just 
as a Deployment."

For per-node scraping, DaemonSets come in handy. "This allows you to scrape the 
kubelet that’s run on every node, it allows you to scrape the node exporter 
that’s also run on every node, which is another Prometheus daemonset that most 

> people run," he expalins.

Author: Wow definitely didn't realize I had so many "expalins", thank you @svrnm 🙈

DaemonSets are useful for scaling out, since they guarantee that you've got pods 
running on every node that matches its selector. "If you have a cluster of 800+ 
nodes, it’s more reliable to run a bunch of little collectors that get those tiny 
metrics, rather than a few bigger stateful set pods because your blast radius is 
much lower," he says. 

"If one pod goes down, you lose just a tiny bit of data, but remember, with all 
this cardinality stuff, that’s a lot of memory. So if you’re doing a StatefulSet, 
scraping all these nodes, that’s a lot of targets, that’s a lot of memory, it 
can go down much more easily and you can lose more data." 

> If a Collector goes down, it comes back up quickly, since it is stateless, which 

Reviewer 2: ```suggestion
If a Collector goes down, it comes back up quickly, since it is usually stateless, which 
```

[Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/), 
[DaemonSet](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/), 
[StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/), 
and [Sidecar](). Which ones you should use depends on what you need to do, such 
as how you like to run applications for reliability. 

"Sidecar is the one we use the least and is probably used the least across the 
industry if I had to make a bet," Jacob says. "They’re expensive. If you don’t 
really need them, then you shouldn’t use them." An example of something run as a 
sidecar is Istio, "which makes a lot of sense to run as a sidecar because it 
does proxy traffic and it hooks into your container network to change how it 
all does its thing."

You will get a cost hit if you sidecar your Collectors for all your services, 
and you also have limited capabilities. He says, "If you’re making Kubernetes 
APIs calls or attribute enrichment, that’s the thing that would get exponentially 
expensive if you’re running as a sidecar." He shares an example: "...if you have 

> sidecar on 10k pods, then that’s 10k API calls made to the K8s API. That's 

Reviewer 3: ```suggestion
sidecar [Collector using the [k8sattributesprocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/k8sattributesprocessor)] on 10k pods, then that’s 10k API calls made to the K8s API. That's 
```

---
title: 'End-User Q&A Series: Migrating to OTel at Lightstep'
linkTitle: 'End-User Q&A: Migrating to OTel at Lightstep'
date: 2023-07-14
author: '[Reese Lee](https://github.com/reese-lee) (New Relic)'
body_class: otel-with-contributions-from

> spelling: cSpell:ignore Aronoff codepath Boten fluentbit autoscaler kubelet

Reviewer: ```suggestion
spelling: cSpell:ignore Aronoff codepath Boten fluentbit autoscaler kubelet
spelling: cSpell:ignore spanmetrics k8sattributesprocessor
```

library authors and maintainers are always working on new performance features
and improvements to the software.

"It makes migration easy as well. Trying to migrate from an early version of
something to the latest version of something, you miss a lot of breaking changes
potentially, and you have to be careful of that," he says.

He recommends using Dependabot, which they use in OTel. OTel packages update in
lockstep, which means you have to update "a fair amount of packages at once, but
it does do it all for you, which is nice," he says. However, you should be doing
this with all your dependencies, as "CVEs happen in the industry constantly. If
you're not staying up to date with vulnerability fixes then you’re opening
yourself up to security attacks, which you don’t want. 'Do something about it'
is my recommendation."

## Additional Resources

- Catch this conversation in full on the
  [OTel YouTube Channel](https://youtu.be/dpXhgZL9tzU)
- To learn more about the OTel Operator, reach out on
  [CNCF Slack](https://communityinviter.com/apps/cloud-native/cncf) in the
  [#OTel-operator channel](https://cloud-native.slack.com/archives/C033BJ8BASU)

> - Jacob will be back to speak with the EUWG at

Reviewer: ```suggestion
- Jacob will be back to speak with the end-user working group at
```
